# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def addTwo(x):   #
    return x + 2

print(addTwo(4))

# Lambda Function Syntax
# Example of calling a lambda function without a variable
print((lambda x: x + 2)(4)) #double parentheses to call the lamda   

# a function into a function

6
6


#### Saving to a Variable

In [4]:
f_test = lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [5]:
# Multiple Inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))  #x=3. y=5  z=8

# Multiple Inputs with a variable attached
x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))   #printing the x_test then calling the assigning the variables their numbers

120
120


#### Passing a Lambda into a Function

In [7]:
def multiply(f,number):   #making this a function by defining it
    """
        f expects a lambda function
        number expects a number
    """
    return f(number)   
multiply(lambda x: x * x, 4) # lambda 4: 4 * 4   since still inside a comma sets up for the number 4

16

#### Returning a Lambda from a Function

In [9]:
# Regular defined function
def multiply_test(num):
    return num * 4

# Function within a function
def returnFunc():
    test = 4  #showing that you can do work within the function
    
    def multiply(num):
        return num * 2
    return multiply  # only returning this function on line 8
#most likely better to 
f_return = returnFunc()
print(returnFunc())  #this is a funkier version
print(f_return(4))

# Lambda function returned from a regular function
def returnLam(b,c):    #this is the function        the variables are b and c
    return lambda x,a: (x + a) / (b + c) #returning this function  now we are using more variables having X and A
r_lamb = returnLam(2,1)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x1084a6e50>
8
3.3333333333333335


#### If Statements within Lambdas

In [10]:
# Lambda x : True if (condition) else False
f_condition = lambda num : num * 2 if num > 10 else num + 2  # saving the function to a variable
#             =8 then 12 then 10 
print(f_condition(8))

print(f_condition(12))

print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [12]:
# f = lambda x:
# print(lambda **3)  my attempt

f = lambda cube: cube ** 3 #short hand way of delcaring a function

print(f(2))    #calling the function F which is equal to lambda then saying lambda equals 2
print(f(3))
print(f(4))

8
27
64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [13]:
# map(func,iterable(list,dict,tuple,set))   
# Normally the usage of map happens with a pre-defined function -- but we can use lambdas as well
def squared(num,num2):  # mapping over 2 lists
    if num < 10 and num2 < 10:    #giving it a condition
        return num ** 2, num2 ** 2   # we want them turned to the second power
    else:
        return num,num2  # if theyre over 10 it just print the numbers
    
numbers = [4,11,20,3,15,20]  #creating a couple lists for examples   
more_nums = [4,10,3,2,6]  # 2nd list our and our lists are not the same length

squared_nums_map = list(map(squared,numbers,more_nums))  # naming our map then giving it a function
print(squared_nums_map)  # NUMBERS HIGHER THAN 10 ARE RETURNED AS NORMAL WITH ELSE CONDITION ON LINE 6

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [14]:
# map(lambda x: x + 2, list)

# using lambda in map happens in line (or in one line) usually

# list(map(lambda x : x * 2,nums))
                # making a list then wrapping a map
squared_nums_lamb = list(map(lambda x,y : (x ** 2,y ** 2) if x < 10 and y < 10 else (x,y),numbers,more_nums))
print(squared_nums_lamb)

#this is the more efficient way using a map and lambda

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [19]:
double_num = list(map(lambda x,y: (((x * 2) - 1),((y * 2) - 1)) if x > 0 and y > 0 else (x,y), numbers,more_nums))
print(double_num)
# using multiple lists numbers and more_nums
numbers2 = list(map(lambda a,b: (((a * 3) - 1),((b * 3) - 1)) if a > 0 and b > 0 else (a,b), numbers,more_nums))
print(numbers2)

L = lambda x: (x*2) - 1  #christophers function
list_a = [2, 2, 3, 4]

print(list(map(L, list_a)))

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]
[(11, 11), (32, 29), (59, 8), (8, 5), (44, 17)]
[3, 3, 5, 7]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [20]:
names = ['Bob','Andy','Max','Evan','Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
    
    
new_names_list = list(filter(a_names, names)) #wrapping the list
print(new_names_list) #will only print andy and angelica

print(a_names)

['Andy', 'Angelica']
<function a_names at 0x1084b30d0>


#### Using Lambda's with Filter()

In [23]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))  #prints A names
print(new_names_lamb)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [29]:
from statistics import mean #importing statistics list of functions  and imports the mean(mean of numbers)

nums = [2,7,4.2,1.6,9,4.4,4.9]
print(mean(nums)) # prints the mean of this list
# this is the mean 4.728571428571429

avg = mean(nums) #assigning a variable to the mean of the nums and calling it avg
print(avg)  #print it for example but not needed

def mean_test(name): #making it into a function calling it mean_test making a list(name)
    if name < avg:  # setting conditons if its lower than the average
        return True
    else:
        return False
    
new_calc = list(filter(mean_test,nums)) #variable of new_calc and assigning it to 
print(new_calc)

# Lambda function solution
below_avg = list(filter(lambda x: True if x <= mean(nums) else False, nums))
print(below_avg) # the more efficient way using lambda 


above_mean = list(filter(lambda number:True if number >= mean(nums) else False, nums)) #jons solution
print(above_mean) # returns the numbers above mean

4.728571428571429
4.728571428571429
[2, 4.2, 1.6, 4.4]
[2, 4.2, 1.6, 4.4]
[7, 9, 4.9]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [30]:
from functools import reduce #could also just use a for loop

# reduce(function,iterable)

list_1 = [2,4,6,8,10] # making a list

def addNums(num1,num2): #num1 refers to
    return num1 + num2

result_add = reduce(addNums,list_1)  #reducing 

print(result_add)

# Subtract a list of numbers
def subtractNums(num1,num2):
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub) #weve been subtracting every number in the list

30
-26


#### Using Lambda's with Reduce()

In [31]:
#  doing everything above with a lambda now 

result_lamb = reduce(lambda x,y: x + y, list_1)
print(result_lamb)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [35]:
list_example = [1,2,3,4,5,]

# from functools import reduce #  not necessary since we already imported from above but showing for example

lambda_example = reduce(lambda x,y: x * y, list_example)
print(lambda_example)

120


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [36]:
#imagine a russian doll going from bigger to smaller to smaller

def addNums(num): #turning it into its own function 
    # Set base case for recusive function
    if num <= 1:   # if the number is less than or equal to 1 
        print("addNums(1) = 1")  
        return num # and just return the number
    #recursive part of the function
    else:
        print(f"addNums({num}) = {num} + addNums({num - 1})") #formatting string so we can see out values
        return num + addNums(num - 1) #should recursively lower
            # starting with 5 and going down
addNums(5) #running 

addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [37]:
# 5 Factorial which is the same as 5 * 4 * 3 * 2 * 1

def factorial(num):  # better visualized on python tutor
    if num <= 1:    # the factorial will be in limbo until at least one is true
        return 1
    else:
        return num * factorial(num - 1)
    
factorial(5)  # 5 then 4 then 3 then 2 would loop until it reaches 1 then they would all execute going back up

120

In [ ]:
#  just like you need a way to break your loop, you need a way to reach your base case

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [40]:
# return num - subNums(num - 1)
# subtract(10)
# def subNums(num):
#     if num <= 1:
#         print("subNums(1) = 1")
#         return num
#     else:
#         print(f"subNums({num}) = {num} - subNums({num - 1})")
#         return num - subNums(num - 1)
    
# subNums(5)

def subNums (num):  #creating a function to subtract and calling it subNums (num) num is the variable
    if num <= 1:  #if the number is less than or equal 1
        print("subNums(1) = 1") #prints it each time the loop is performing, unessacary but able to visualize
        return num
    else:   #to break the loop we make an else statement
        print(f"subNums{num}) = {num} - subNums({ num - 1})")
        return num - subNums(num - 1)  # breaks the loop 
    
subNums(5)  #calling the function from line 13 and assigning 5 to the variable then making the loop going down

subNums5) = 5 - subNums(4)
subNums4) = 4 - subNums(3)
subNums3) = 3 - subNums(2)
subNums2) = 2 - subNums(1)
subNums(1) = 1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [44]:
def my_range(stop, start, step = 2): #creating a function
    while start < stop:  
        yield start # yield keyword denotes a generator
        start += step
for i in my_range(20, start = 2): # setting my stop at 20 and start at 2
    my_generator_value = i  
    print(my_generator_value)

my_range(20,start = 2)

2
4
6
8
10
12
14
16
18


<generator object my_range at 0x1084e4ba0>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [45]:
# def example_range(stop, start step = 3):
#     while start >= stop:
#         yield start **2 # this squares the number
#         start -= step #  saying we want to go down in our steps
        
# for i in example_range()
        
def num_sq(stop,start,step = 1):
    while start >= stop:
        yield start ** 2
        start -= step
        
for i in num_sq(0,start = 3):
    print(i)
    
def neg_range(stop, start, step = 1):  #matas' example
    while start > stop:
        yield start
        start -= step
        
for i in neg_range(0, start = 20):
    my_neg_gen = i ** 2
    print(my_neg_gen)
    
print(neg_range(0,20))

9
4
1
0
400
361
324
289
256
225
196
169
144
121
100
81
64
49
36
25
16
9
4
1
<generator object neg_range at 0x1084ce430>


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [2]:
import re

places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

# pattern = re.compile("([A-Z][A-Za-z]+) ([A-Za-z].*)") 

pattern_1 = re.compile('[\w]+')

places = pattern_1.findall

print(places)

places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

def a_places(place):
    if place == "Argentina":
        return True
    elif place == "San Diego":
        return True
    elif place == "Boston":
        return True
    elif place == "New York":
        return True
    else:
        return False
    
new_place_list = list(filter(a_places,places))
print(new_place_list)


<built-in method findall of re.Pattern object at 0x110655850>
['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [4]:


authors = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda writer: writer.split(" ")[-1].lower())

print(authors)

['Joel Carter', 'Victor aNisimov', 'Andrew P. Garfield', 'David hassELHOFF', 'Gary A.J. Bernstein']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [5]:
# F = (9/5)*C + 32
# places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

# print(map(places)))


L = lambda x: (x*2) - 1  #christopher helped me with this one
list_a = [2, 2, 3, 4]

print(list(map(L, list_a)))


[3, 3, 5, 7]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [7]:
def fib(i):
    if i <= 1:
        return 1
    else:
        return fib(i-1)+fib(i-2)
    
fibo(7)

21